### PR TITLE
feat(cloudfront): `function URL` origin access control L2 construct

### PR DESCRIPTION
**Pull request by watany-dev**
Issue # (if applicable)

#31629 

### Reason for this change
This change introduces support for Lambda Function URLs with custom Origin Access Control (OAC) in CloudFront distributions, enhancing security and control over CloudFront-Lambda integration.

### Description of changes
- Added a new feature allowing the configuration of Lambda Function URLs with custom OAC in CloudFront.
- Implemented support for custom signing behavior and protocols for Lambda origins.
- Included new tests to validate the correct behavior of OAC with Lambda Function URLs.

### Description of how you validated changes
- Ran unit tests to ensure that the OAC setup for Lambda Function URLs is correctly applied in CloudFront distributions.
- Validated by deploying a sample CDK application to confirm the functionality and integration of Lambda Function URLs with CloudFront using OAC.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)
- [x] OAC implementation is complete.
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-cloudfront/test/origin-access-control.test.ts
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-cloudfront/test/origin-access-control.test.ts:63]: Can we nest the new tests under a new `describe`, e.g. 
```
describe('FunctionUrlOriginAccessControl', () => {
    test(...);
});
```

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts:88]: why is `undefined` passed here? Could we just leave out the props altogether?

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts:115]: Can we move this into a private function? It would make it clearer and easier to understand why a new permission is getting created

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (gracelu0) [inline packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:30]: Can we add an assertion to this test that calls the cloudfront distribution endpoint and checks the response is as expected?

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts:95]: since this test is checking for the correct permissions, can we also check the `AWS::Lambda::Permission` resource exists in the template and has the expected policy?

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (gracelu0) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts:115]: Also is it possible to use any existing grant or addPermission methods here instead? And how do we handle permissions for imported lambda functions?

**Review by gracelu0 — CHANGES_REQUESTED** [review]
Thanks for working on this so quickly @watany-dev! :)

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts]: It seems like there might still be a potential bug with addPermission, so I’m using cfnPermission instead. It’s not too complex, and since the cross-stack test is passing, I don't have any concerns. Here’s a reference (apologies, it's in Japanese!) https://dev.classmethod.jp/articles/cdk-over-21-lambda-create-error/.

**Comment by watany-dev** [comment]
note: There are still remaining fixes needed for the integration tests.

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts]: fixed. 
https://github.com/aws/aws-cdk/pull/31339/commits/a43bf40728eb7ed5bf1fb25afa3a5ca56ccca88e

File: packages/aws-cdk-lib/aws-cloudfront/test/origin-access-control.test.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront/test/origin-access-control.test.ts:63]: fixed.
https://github.com/aws/aws-cdk/pull/31339/commits/3731370b9b8f8f6362f6d9b4b983108767f0dff2

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts]: fixed.
https://github.com/aws/aws-cdk/pull/31339/commits/f660a837b42cfac56fb2a5439b4aa661523ccaaf

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts]: fixed.
https://github.com/aws/aws-cdk/pull/31339/commits/2ad91705267267ed03c9ab85a1b54d96c7e6bf80

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts]: fixed.
https://github.com/aws/aws-cdk/pull/31339/commits/7226680a2e61894796626e5e23b54f5d9586490d

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (watany-dev) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:30]: fixed.

**Comment by watany-dev** [comment]
@gracelu0 
Thank you for the swift review. All tasks, including the integration test, have been completed. Could you please verify?

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (gracelu0) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:30]: Thanks for adding those assertions! What I was thinking of originally was using the `httpApiCall` method to the distribution domain and verifying the response from the lambda (something like [this test](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.s3-encrypted-bucket-origin-oac.ts#L36-L39))

```
    const resp = JSON.stringify({
      statusCode: 200,
      body: 'Hello!!'
    });

    const fn = new lambda.Function(this, 'MyLambda', {
      runtime: lambda.Runtime.NODEJS_20_X,
      code: lambda.Code.fromInline(`exports.handler = async (event) => {return ${resp}}`),
      handler: 'index.handler'
    });
    ...
    const call = integ.assertions.httpApiCall(`https://${distribution.distributionDomainName}`);
    call.expect(ExpectedResult.objectLike({
      status: 200,
      body: 'Hello!!',
    }));
```
Let me know if that wouldn't work for some reason.

Comment (watany-dev) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:35]: @gracelu0 

Your comment passed the test correctly. This will ensure its appropriateness.

**Comment by watany-dev** [comment]
@gracelu0 

Sorry for rushing you when you're busy, but could I ask you to review this?

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (godwingrs22) [inline packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts:146]: can we refactor the test case name to be more specific to the test case?
```
For eg: Creates Lambda Function URL origin with default Origin Access Control
```

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (godwingrs22) [inline packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts:159]: can we also have a test case that accepts only fnUrl?

```
FunctionUrlOrigin.withOriginAccessControl(fnUrl)
```

Comment (godwingrs22) [reply packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts:159]: Also a test case for OAC with default origin properties and with customOriginConfig properties
```
originSslProtocols
originProtocolPolicy
originReadTimeout
originKeepaliveTimeout
```

**Review by godwingrs22 — DISMISSED** [review]
Thanks @watany-dev for the great work. Left some minor comments.

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts]: fixed. https://github.com/aws/aws-cdk/pull/31339/commits/258e464d68015451c8c38adef0fc8f8894f652a9

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts:107]: fixed. https://github.com/aws/aws-cdk/pull/31339/commits/16eb2274d8cc79d592869b79972871f9ecb4b0ae

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts]: fixed. https://github.com/aws/aws-cdk/pull/31339/commits/d8b0b1669c37541c79acb94c8349991bc488e79c

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts:159]: fixed https://github.com/aws/aws-cdk/pull/31339/commits/d8b0b1669c37541c79acb94c8349991bc488e79c

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (godwingrs22) [inline packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:25]: Just for my understanding, I would like to understand the behaviour of `AuthType` in Function URL with Cloudfront-Lambda integration with OAC origin.

1. May i know if `authType = AWS_IAM` is the only supported lamda function URL for Cloudfront OAC Lambda Function url origin?
2. Or Is `authType = None` supported for OAC? If it is not supported, Let us mention it in ReadMe and also update the current example.
3. If both are supported, I would recommend to have unit test for both types.

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (godwingrs22) [inline packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:31]: Also I would recommend to have a another integ test for the custom OAC with new L2 construct `FunctionUrlOriginAccessControl` and using it in `withOriginAccessControl` method

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (godwingrs22) [inline packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts:95]: I believe we need `this.originAccessControl = this.props?.originAccessControl` still to set the custom OAC. 

This is because `originAccessControlId` is being used in the `bind` method from this.originAccessControl. Hence for scenario where custom OAC is passed, I suspect it wont be able to retrieve the originAccessControlId based on the current logic. Please correct me if my understanding is incorrect.

Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts:97]: You are right. I fixed it.

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (watany-dev) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:25]: OAC can communicate when authType = none and the necessary permissions are granted. When None is set in the L2 construct, the latter permission is automatically configured, so there are no issues. Therefore, I have prepared a test for the None case.

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (watany-dev) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:40]: fixed,

**Review by godwingrs22 — DISMISSED** [review]
Thanks @watany-dev for addressing the feedback. I left few more comments. Also, Please be informed that this PR requires security review and I'm checking internally on the same. I'll let you know once it is security approved.

**Comment by watany-dev** [comment]
@godwingrs22 

I understand the content, but this cannot be achieved with the current implementation.

The lambda.FunctionUrl property does not include authType, which means, for example, the URL Auth Type of an imported function cannot be identified within the Origin construct.

May I request permission to add this property as a prerequisite patch? Once this property is added, we can merge it into the main branch, either in this commit or a separate one, to resolve this issue.

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (watany-dev) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts]: fixed
https://github.com/aws/aws-cdk/pull/31339/commits/fd9f9b45700e8db27aa5740a8466927d4d2b875d

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (watany-dev) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts]: fixed. https://github.com/aws/aws-cdk/pull/31339/commits/873777b1de8239a347f45a896792c1fe43ee2b15

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts]: fixed, https://github.com/aws/aws-cdk/pull/31339/commits/fd9f9b45700e8db27aa5740a8466927d4d2b875d

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (godwingrs22) [inline packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts:376]: nit: seems this test is duplicated. It is similar to test `Correctly creates a Lambda Function URL Origin with default properties`

**Comment by godwingrs22** [comment]
> The lambda.FunctionUrl property does not include authType, which means, for example, the URL Auth Type of an imported function cannot be identified within the Origin construct.

I see. Thanks for checking.

> May I request permission to add this property as a prerequisite patch? Once this property is added, we can merge it into the main branch, either in this commit or a separate one, to resolve this issue.

Sorry, I didn't get it. Could you please elaborate your suggestion?

I think we can also update the readMe to specify to use the recommended value for Lambda Function url AuthType as `AWS_IAM` if OAC Signing `SIGV4_ALWAYS` is used. Also update the authtype given in the readMe example to `AWS_IAM` instead of `None`.

File: packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/test/function-url-origin.test.ts]: https://github.com/aws/aws-cdk/pull/31339/commits/2375bcd40c2e7512d1cf0cad40dd6bf98b09d3a7

**Comment by watany-dev** [comment]
@godwingrs22 
I assumed that I was being asked for a validation feature, so I submitted a pull request at https://github.com/aws/aws-cdk/pull/31590. However, if this is unnecessary at this point and can be resolved by updating the Readme first, how does this draft look?
https://github.com/aws/aws-cdk/pull/31339/commits/fcf8db0727025ff89ca22e33ee33511f999dcef8

**Comment by godwingrs22** [comment]
Thank you @watany-dev for addressing the feedback promptly and appreciate your great effort.

> I assumed that I was being asked for a validation feature, so I submitted a pull request at #31590. However, if this is unnecessary at this point and can be resolved by updating the Readme first, how does this draft look?

Let us keep the current readme with recommendation(which you have already updated) and also have this validation feature as well for the recommended AWS best practises. Hence, let's get the other PR #31590 merged first and then we can add this validation later in this PR. Regarding the validation, let's throw synth warning instead of exception to use the recommended value for function url `authType = AWS_IAM` if OAC signing is `SIGV4_ALWAYS`

Aside, I'm also following up internally for the security review and will let you know incase if there is any changes needs to be made from security prespective.

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts:121]: Thanks for adding the validation as requested @watany-dev ! Could we rename the method to `validateAuthType` for consistency? We are still getting this PR security reviewed as well, so we may ask to throw an error instead of warning here for best security practices. But I will confirm first so you can hold off on that change for now.

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts:153]: @gracelu0 
Sorry, I couldn't think of a concrete way to do it here, it can't be referenced from L2 Construct and the usual `this.something.node.defaultChild as ` doesn't reference this property!

**Comment by rhermes62** [comment]
Just to confirm, will this proposal work well with Lambda function aliases and versions?

Background: Lambda [provisioned concurrency](https://docs.aws.amazon.com/lambda/latest/dg/provisioned-concurrency.html) only works with Lambda function versions or aliases. Given that, we have all of our function URLs pointing to aliases instead of "normal" Lambda functions. Before this was merged, just wanted to make sure this PR would also accommodate function URLs backed by versions/aliases.

From updated README looks like the answer is yes and it would be something like:

```
const alias = fn.addAlias('Live', { provisionedConcurrentExecutions });

const aliasUrl = alias.addFunctionUrl({
  authType: lambda.FunctionUrlAuthType.AWS_IAM,
});

new cloudfront.Distribution(this, 'MyDistribution', {
  defaultBehavior: {
    origin: origins.FunctionUrlOrigin.withOriginAccessControl(aliasUrl),
  },
});
```

Might be worth adding to README updates given this may be a pretty common pattern.

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac-custom.ts
Comment (gracelu0) [inline packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac-custom.ts:104]: nit: we no longer need `app.synth()` in integ tests so this line can be removed from the new integ tests

**Comment by gracelu0** [comment]
> Just to confirm, will this proposal work well with Lambda function aliases and versions?
> 
> Background: Lambda [provisioned concurrency](https://docs.aws.amazon.com/lambda/latest/dg/provisioned-concurrency.html) only works with Lambda function versions or aliases. Given that, we have all of our function URLs pointing to aliases instead of "normal" Lambda functions. Before this was merged, just wanted to make sure this PR would also accommodate function URLs backed by versions/aliases.
> 
> From updated README looks like the answer is yes and it would be something like:
> 
> ```
> const alias = fn.addAlias('Live', { provisionedConcurrentExecutions });
> 
> const aliasUrl = alias.addFunctionUrl({
>   authType: lambda.FunctionUrlAuthType.AWS_IAM,
> });
> 
> new cloudfront.Distribution(this, 'MyDistribution', {
>   defaultBehavior: {
>     origin: origins.FunctionUrlOrigin.withOriginAccessControl(aliasUrl),
>   },
> });
> ```
> 
> Might be worth adding to README updates given this may be a pretty common pattern.

Thanks for pointing out this use case! I tested setting up a Lambda function alias with OAC with this API and it deployed/setup successfully. @watany-dev could you please add an integration test and README example for this use case?

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac-custom.ts
Comment (watany-dev) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac-custom.ts]: fixed.
https://github.com/aws/aws-cdk/pull/31339/commits/baddb09fbefe093e483f909d721faf8841231fa9

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts]: fixed. https://github.com/aws/aws-cdk/pull/31339/commits/f95862117efce6ccf185b7f320f294f5005a2802

**Comment by watany-dev** [comment]
@rhermes62 @gracelu0 
Thank you. I am happy to share that I was also able to confirm the Alias pattern.
https://github.com/aws/aws-cdk/pull/31339/commits/8be1369ce8462835cdecf0c176d467ecc13a9e2f

**Comment by gracelu0** [comment]
Thank you @watany-dev ! We are still pending the security review and service team confirmation, we appreciate your effort and patience in getting this reviewed.

File: packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts
Comment (ViswanathaReddyGajjala) [reply packages/@aws-cdk-testing/framework-integ/test/aws-cloudfront-origins/test/integ.function-url-origin-oac.ts:25]: I agree with @godwingrs22 . It's appropriate to throw an exception if `authType` is `None` and `OAC` Signing is `SIGV4_ALWAYS`. This configuration leads to errors and can be confusing for the customer.

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (gracelu0) [inline packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts:122]: Confirmed that we should throw an error here instead of warning to enforce security best practice. Could you please update the error message as well to `The authType of the Function URL must be set to AWS_IAM when origin access control signing method is SIGV4_ALWAYS.`?

**Review by gracelu0 — DISMISSED** [review]

File: packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts
Comment (watany-dev) [reply packages/aws-cdk-lib/aws-cloudfront-origins/lib/function-url-origin.ts]: This commit addresses the following changes:
https://github.com/aws/aws-cdk/pull/31339/commits/86f60d85c7d4f2ccc376aa45edb1c3e292901d32

- Modification from Warning to Error: This change updates the handling from a warning to an error, which corresponds to the comment in question.
- Detailed Escape Patch for cfnOriginAccessControl: This improves handling of unknown specifications, enhancing type safety.
- Error Handling within Validation Methods: Error handling is now managed within validation methods, reducing complex value passing between classes.

**Comment by watany-dev** [comment]
@gracelu0 
Could you please provide an update on the status of this review?

**Comment by gracelu0** [comment]
Hi @watany-dev , I think it looks good! Just waiting on confirmation from security reviewer before I approve (should be very soon!).

**Review by gracelu0 — APPROVED** [review]
Thank you so much for working on this @watany-dev !

**Pull request merged**
